### ISSUE TITLE
Do the last FC layers get out of the Hyperbolic space?

### ISSUE THREAD

**Issue by dragon9001** [bcol23/HyperIM#3]
The network architecture begins with a few hyperbolic layers but ends with Euclidean layers as states in:
https://github.com/bcol23/HyperIM/blob/c257d1c6296e399e734110d54d0ccd5fc91b5c02/net/HyperIM.py#L32

The full architecture was:

``` python
    def __init__(self, feature_num, word_embed, label_embed, hidden_size=5, if_gru=True, 
                 default_dtype=th.float64, **kwargs):
        super().__init__(**kwargs)
        
        self.hidden_size = hidden_size
        
        self.word_embed = gt.ManifoldParameter(word_embed, manifold=gt.PoincareBall())
        self.label_embed = gt.ManifoldParameter(label_embed, manifold=gt.PoincareBall())
        self.default_dtype = default_dtype
        
        if(if_gru):
            self.rnn = hyperGRU(input_size=word_embed.shape[1], hidden_size=self.hidden_size, 
                                default_dtype=self.default_dtype)
        else:
            self.rnn = hyperRNN(input_size=word_embed.shape[1], hidden_size=self.hidden_size, 
                                default_dtype=self.default_dtype)
        
        self.dense_1 = nn.Linear(feature_num, int(feature_num/2))
        self.dense_2 = nn.Linear(int(feature_num/2), 1)
```

How can we optimize with Riemann SGD if not all the parameters are on hyperbolic space?

**Comment by bcol23** [comment]
The final MLP uses the similarity scores to do the prediction part so its parameters are in the Euclidean space and are updated via vanille optimization methods.

Only the word embeddings and label embeddings (and the bias in the hyperbolic RNN/GRU) are located in the hyperbolic space and are updated with Riemannian optimization methods.

**Comment by dragon9001** [comment]
Does Riemannian Optimizer distinguish Euclidean parameters with Hyperbolic parameters? 
Because only one optimizer is used as:

https://github.com/bcol23/HyperIM/blob/c257d1c6296e399e734110d54d0ccd5fc91b5c02/HyperIM.py#L41-L45

**Comment by bcol23** [comment]
For parameters in the hyperbolic manifold, RiemannianAdam is applied, while vanilla Adam is applied to other parameters in the Euclidean space.
